python/whylogs/experimental/core/udf_schema.py

```diff
@@ -0,0 +1,89 @@
+import logging
+from copy import deepcopy
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, List, Mapping, Optional, Tuple
+
+from whylogs.core.datatypes import DataType, TypeMapper
+from whylogs.core.metrics.metrics import MetricConfig
+from whylogs.core.resolvers import ResolverSpec
+from whylogs.core.schema import DatasetSchema, DeclarativeSchema
+from whylogs.core.segmentation_partition import SegmentationPartition
+from whylogs.core.stubs import pd
+from whylogs.core.validators.validator import Validator
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class UdfSpec:
+    column_name: Optional[str] = None  # TODO: maybe make this a regex
+    column_type: Optional[DataType] = None
+    udfs: Dict[str, Callable[[Any], Any]] = field(
+        default_factory=dict
+    )  # new column name -> callable to compute new column value
+
+
+class UdfSchema(DeclarativeSchema):
+    def __init__(
+        self,
+        resolvers: List[ResolverSpec],
+        types: Optional[Dict[str, Any]] = None,
+        default_config: Optional[MetricConfig] = None,
+        type_mapper: Optional[TypeMapper] = None,
+        cache_size: int = 1024,
+        schema_based_automerge: bool = False,
+        segments: Optional[Dict[str, SegmentationPartition]] = None,
+        validators: Optional[Dict[str, List[Validator]]] = None,
+        udf_specs: Optional[List[UdfSpec]] = None,
+    ) -> None:
+        super().__init__(
+            resolvers=resolvers,
+            types=types,
+            default_config=default_config,
+            type_mapper=type_mapper,
+            cache_size=cache_size,
+            schema_based_automerge=schema_based_automerge,
+            segments=segments,
+            validators=validators,
+        )
+        self.udf_specs = udf_specs or []
+
+    def copy(self) -> DatasetSchema:
+        copy = super().copy()
+        copy.udf_specs = deepcopy(self.udf_specs)
+        return copy
+
+    def _run_udfs_on_row(self, row: Mapping[str, Any], new_columns: Mapping[str, Any]) -> None:
+        for column, value in row.items():
+            for udf_spec in self.udf_specs:
```

Comment:
Change this to map from column name -> UDF list and map from column type -> UDF list for efficiency